core/src/saros/negotiation/AdditionalProjectDataFactory.java

```diff
@@ -0,0 +1,79 @@
+package saros.negotiation;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CopyOnWriteArrayList;
+import org.apache.log4j.Logger;
+import saros.filesystem.IProject;
+
+/**
+ * Factory class to build the additional project data used for the project negotiation data.
+ *
+ * <p>To add entries to the additional project data mapping, a {@link ProjectDataProvider} can be
+ * registered. These project data providers will be called with the corresponding project object
+ * every time project negotiation data is created for a shared project.
+ *
+ * <p>The registered project data providers should provide a way of identifying their registered
+ * options (e.g. by providing references to the used key values).
+ *
+ * <p><b>NOTE:</b> This factory does not ensure that the keys used by different project data
+ * providers are disjoint. If multiple providers use the same key, the mapping of the provider added
+ * last will be used.
+ */
+public class AdditionalProjectDataFactory {
```

Comment:
I would either throw this into the PicoContainer or at least using the singleton pattern for this class.

Reply:
Basically it just the design pattern. Leave it as it is, however the approach is very bad. We should not start overuse static access to some "components" or otherwise we end up in global state programming.

Reply:
I mean I can change it to a singleton as it doesn't really change anything for accessing the class. As already said, I don't really mind. Would this be preferable to you?

Or are you set on having it be part of the context?

Reply:
Having it part of the content would be nice, and I do not think it is that complicated to get it into the IPN / OPN.